### PR TITLE
[Model] Create an RGCN model using cugraph-ops

### PR DESCRIPTION
**Pull request by tingyu66**
## Description

This PR adds an accelerated R-GCN model that uses the aggregation functions in `pylibcugraphops`, a Python wrapper of `cugraph-ops`.
Fixes https://github.com/rapidsai/cugraph-ops/issues/179.

To install `pylibcugraphops`:
```
conda install -c rapidsai-nightly pylibcugraphops
```
(change it to 22.10 after its release.)

The class `dgl.contrib.cugraph.nn.RelGraphConv` uses a pytorch backend and offers similar functionality and interface as [`dgl.nn.RelGraphConv`](https://docs.dgl.ai/en/latest/generated/dgl.nn.pytorch.conv.RelGraphConv.html), with the only difference being:
- `regularizer` has to be either `None` or `'basis'` (not supporting block-diagonal regularization)
- The `__init__()` function of the class take an extra parameter `fanout` as input. This is because the aggregators in `pylibcugraphops` are designed for mini-batch training and *not* suitable for full-graph use cases. Users should pass in the same `fanout` value used for creating DGL's sampler. Consequently, the `forward` function expect input graphs to be a `DGLBlock` (message flow graph, or MFG).
- ~~The node type and edge type tensor of the DGL graph must be of type `torch.int32`, regardless of the node/edge ID type `g.idtype` being `int32` or `int64`.~~

This PR includes an entity classification example `entity_sample.py`. Compared to the original dgl [example](https://github.com/dmlc/dgl/blob/master/examples/pytorch/rgcn/entity_sample.py), users would only need to change the imports and pass in `fanouts` to the model in order to use this cugraphops-accelerated variant. One caveat is that when the `test_sampler` has a different `fanouts` than the `train_sampler`, we have to update the fanouts in the conv layers using `update_fanouts()`.

## Performance
Hardware: 32-core AMD Ryzen Threadripper Pro 3975WX, NVIDIA RTX A6000
Benchmark: Mini-batch `entity_sample.py` example:
- batch_size: 100, graph idtype is int64.
- 2-layer RGCN, with fanouts: (4,4), hidden embedding dimension: 16, num_bases: 20
- The timings are for 100 epochs and the average across 3 identical runs.

|       | time_dgl (s) | time_cugraph-ops (s) |
|:-----:|:------------:|:--------------------:|
|  aifb |     4.24     |         2.61         |
| mutag |     5.65     |         3.68         |
|  bgs  |     4.19     |         2.70         |
|   am  |     19.11    |         13.71        |

During forward aggregation, most of the time (~70%) is spent on the conversion from DGL graph to cugraph-ops MFG type, i.e., `get_csr`:  grab csr representation of the DGL graph and `create_mfg`: create cugraph-ops MFG. The actual forward aggregation function only takes 1/5 of the total time. 

```
Total Time (ns)  Instances   Avg (ns)     Med (ns)    Min (ns)  Max (ns)   StdDev (ns)   Style       Range    
---------------  ---------  -----------  -----------  --------  ---------  -----------  -------  -------------
    834,193,721        814  1,024,808.0  1,015,957.0   957,171  2,264,271     96,833.9  PushPop  total_agg_fwd
    395,471,358        814    485,837.0    474,488.0   453,493  1,577,778     78,678.5  PushPop  create_mfg   
    182,373,749        814    224,046.4    233,108.5   195,217    398,820     22,571.4  PushPop  get_csr      
    159,149,002        814    195,514.7    193,233.5   185,399    488,559     13,795.3  PushPop  agg_fwd_func
```

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### Notes
This PR is meant for tracking progress, as well as internal code review, and should not be merged into the master. When it's ready for external review, we should change the base repository of the PR to DGL's upstream repo.

### CODE REVIEW

**Comment by teju85** [comment]
Cc += @MatthiasKohl  and @stadlmax as well.

**Review by wangxiaoyunNV — APPROVED** [review]

**Comment by MatthiasKohl** [comment]
I have a few comments (they don't need to be addressed in this PR, but either that or we should create issues for these):

1. The major one: right now, we do *not* support any fan-out, the underlying functions are not intended for full-graph use-cases, and sampled use-cases can typically assume a fan-out <100 or so, as well as fan-out staying fixed. This means that we should *not* calculate the fan-out on the fly, this is a very excessive overhead which is useless in sampled cases since we'll know at model construction time how much we want to sample. So the solution here is to add a parameter for the fan-out on `__init__` of `RelGraphConvOps`.
2. In general, the goal should not be to create some sort of replacement for DGL's `RelGraphConv`, the goal should be to provide an alternative for *specific* use-cases which can be seamlessly used: It would be much better IMO if we would directly integrate into DGL's `RelGraphConv` by checking in which cases our methods apply. For now with the first few PRs, it can be done separately of course, especially for testing purposes, but the goal should be to have DGL's API check for which cases apply to our methods and use them there.
3. Node/edge types should allow int64, we can just cast them otherwise. This will be expensive but can be explained in the docs of `RelGraphConvOps`, so the user is advised to provide `int32` directly.
4. I wouldn't include the matrix multiplication in the autograd function. It complicates the backward and it is reasonable to add the bias there directly, allowing much easier fusion. If you do include it, you should at least check for whether gradients are actually necessary (using `ctx.needs_input_grad`).
5. the self loop weight should be fused into the rest of the weights. See above, this can be done with a single linear layer if we remove weights from the autograd function. Care needs to be taken for initializing the weights according to how DGL currently does it, but should be very straightforward. This allows much better automatic fusion.

**Comment by tingyu66** [comment]
@MatthiasKohl  Thanks for the feedback. I will address these concerns in this PR. `RelGraphConvOps` is an intermediate for testing, and its integration into `RelGraphConv` can happen in a separate PR.

File: python/dgl/contrib/cugraph/nn/conv/relgraphconv_ops.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+            grad_coeff = None
+            agg_bwd_func(grad_feat, grad_output, feat.detach(), mfg)
+        else:
+            grad_coeff = th.empty_like(coeff, dtype=th.float32, device='cuda')
+            agg_bwd_func(grad_feat, grad_output, feat.detach(), mfg,
+                output_weight_gradient=grad_coeff, weights_combination=coeff.detach())
+
+        return None, None, None, None, None, None, grad_feat, grad_coeff
+
+class RelGraphConvOps(nn.Module):
```
Comment (VibhuJawa) [inline python/dgl/contrib/cugraph/nn/conv/relgraphconv_ops.py]: Why did we name this class `RelGraphConvOps` and not `RelGraphConv`  and keep it consistent with DGL ? Keeping the naming consistent with upstream will make it a drop in replacement for most examples.

File: python/dgl/contrib/cugraph/nn/conv/entity_sample.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+from time import time
+import torch as th
+import torch.nn as nn
+import torch.nn.functional as F
+from torchmetrics.functional import accuracy
+
+import dgl
+from dgl.data.rdf import AIFBDataset, MUTAGDataset, BGSDataset, AMDataset
+from dgl.dataloading import MultiLayerNeighborSampler, DataLoader
+from dgl.contrib.cugraph.nn import RelGraphConvOps
```
Comment (VibhuJawa) [inline python/dgl/contrib/cugraph/nn/conv/entity_sample.py]: I think we should make it a point to show that only the import for `dgl.contrib.cugraph.nn`  changes and nothing else to make it clear to the users.

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
Have requested a rename to match DGL but looks good otherwise as well as clarification from a user perspective.

File: python/dgl/contrib/cugraph/nn/conv/relgraphconv_ops.py (hunk 1 above)
Comment (tingyu66) [reply python/dgl/contrib/cugraph/nn/conv/relgraphconv_ops.py]: You are right. Fixed in [6c34abc](https://github.com/rapidsai/dgl/pull/15/commits/6c34abc900beaa6231354b970f2318210f397634).

File: python/dgl/contrib/cugraph/nn/conv/entity_sample.py (hunk 2 above)
Comment (tingyu66) [reply python/dgl/contrib/cugraph/nn/conv/entity_sample.py]: [6a95973](https://github.com/rapidsai/dgl/pull/15/commits/6a95973a122ee769096329df2c35ef258d31fe38) fixed the imports to match the DGL example.

**Review by VibhuJawa — APPROVED** [review]
LGTM

File: python/dgl/contrib/cugraph/nn/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+from .conv import *
```
Comment (VibhuJawa) [inline python/dgl/contrib/cugraph/nn/__init__.py]: We will  have to make this optional for upstreaming. 
```suggestion
try:
     from .conv import *
except ModuleNotFoundError:
        raise ModuleNotFoundError("dgl.contrib.cugraph.nn requires pylibcugraphops to be installed")
```

File: python/dgl/contrib/cugraph/nn/conv/entity_sample.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
+import torch
```
Comment (VibhuJawa) [inline python/dgl/contrib/cugraph/nn/conv/entity_sample.py]: We probably need to move this to the examples directory. 

https://github.com/dmlc/dgl/tree/master/examples

File: python/dgl/contrib/cugraph/nn/conv/entity_sample.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='RGCN for entity classification with sampling')
+    parser.add_argument("--dataset", type=str, default="aifb",
+                        help="Dataset name ('aifb', 'mutag', 'bgs', 'am'), default to 'aifb'.")
+    parser.add_argument("--idtype", type=str, default="int64",
+                        help="Graph idtype ('int64' or 'int32'), default to 'int64'.")
+    args = parser.parse_args()
+    device = torch.device('cuda')
+    print(f'Training with DGL built-in RGCN module with sampling.')
```
Comment (VibhuJawa) [inline python/dgl/contrib/cugraph/nn/conv/entity_sample.py]:
```suggestion
    print(f'Training with DGL cugraph.nn RGCN module with sampling.')
```

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
Thanks for adding the test cases. 

Have requested minor changes .  We also need to add the API documentation to below . 

https://github.com/dmlc/dgl/tree/master/docs/source/api/python


We can probably follow https://docs.dgl.ai/en/0.8.x/api/python/dgl.contrib.UnifiedTensor.html as an example

**Review by VibhuJawa — CHANGES_REQUESTED** [review]
LGTM

**Review by VibhuJawa — APPROVED** [review]
LGTM, thanks again for making these changes @tingyu66  . Looks good to upstream to me.
